service/dynamodb/expression/expression.go

```diff
@@ -0,0 +1,630 @@
+package expression
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+)
+
+// ErrEmptyFactoryBuilder is an error that is returned if BuildFactory() is
+// called on an empty FactoryBuilder.
+var ErrEmptyFactoryBuilder = awserr.New("EmptyFactoryBuilder", "BuildFactory error: the argument FactoryBuilder is empty", nil)
+
+// expressionType will specify the type of Expression. The const is used to
+// eliminate magic strings
+type expressionType string
+
+const (
+	projection   expressionType = "projection"
+	keyCondition                = "keyCondition"
+	condition                   = "condition"
+	filter                      = "filter"
+	update                      = "update"
+)
+
+// Implementing the Sort interface
+type typeList []expressionType
+
+func (list typeList) Len() int {
+	return len(list)
+}
+
+func (list typeList) Less(i, j int) bool {
+	return string(list[i]) < string(list[j])
+}
+
+func (list typeList) Swap(i, j int) {
+	list[i], list[j] = list[j], list[i]
+}
+
+// FactoryBuilder contains a list of structs fulfilling the TreeBuilder
+// interface, representing the different types of Expressions that make up the
+// total Expression as a whole. FactoryBuilder will have methods corresponding
+// to different types of expressions (Update(), Condition(), Filter(), etc) that
+// will allow users to add expressions to the input member Factory.
+// FactoryBuilder will have a method Build() which will build the Factory struct
+// which can be used to produce members of DynamoDB input structs.
+//
+// Example:
+//
+//     factoryBuilder := expression.KeyCondition(
+//       expression.Key("someKey").Equal(expression.Value("someValue"))
+//     ).Projection(
+//       expression.NamesList("aName", "anotherName", "oneOtherName")
+//     )
+//     factory := factoryBuilder.BuildFactory()
+//
+//     queryInput := dynamodb.QueryInput{
+//       KeyConditionExpression:    factory.KeyCondition(),
+//       ProjectionExpression:      factory.Projection(),
+//       ExpressionAttributeNames:  factory.Names(),
+//       ExpressionAttributeValues: factory.Values(),
+//       TableName: aws.String("SomeTable"),
+//     }
+type FactoryBuilder struct {
```

Comment:
Suggest instead of a factor builder being constructed by things like `Condition` switch it around and define an `ExpressionBuilder` which as helper methods to add `Condition`, `Update` ect.

```go
type Builder struct {}
func NewBuilder() Builder {}

func (e Builder) WithCondition(builder ConditionBuilder) ExpressionBuilder {}
func (e Builder) WithUpdate(builder UpdateBuilder) ExpressionBuilder {}
...
func (e Builder) Build() Expression {}

type Expression struct {}
func (e Expression) Names() map[string]*string {}
func (e Expression) Values() map[string]*dynamodb.AttributeValue {}

func (e Expression) Get(expType Type) *string {}
// or 
func (e Expression) Condition() *string {}
```